libs/langchain/langchain/vectorstores/redis/base.py

```diff
@@ -0,0 +1,1351 @@
+
+    Raises:
+        ValueError: If a metadata field cannot be categorized into any of
+            the three known types.
+    """
+    result: Dict[str, Any] = {
+        "text": [],
+        "numeric": [],
+        "tag": [],
+    }
+
+    for key, value in data.items():
+        # Numeric fields
+        try:
+            int(value)
+            result["numeric"].append({"name": key})
+            continue
+        except (ValueError, TypeError):
+            pass
+
+        # None values are not indexed as of now
+        if value is None:
+            continue
+
+        # if it's a list of strings, we assume it's a tag
+        if isinstance(value, (list, tuple)):
+            if not value or isinstance(value[0], str):
+                result["tag"].append({"name": key})
+            else:
+                name = type(value[0]).__name__
+                raise ValueError(
+                    f"List/tuple values should contain strings: '{key}': {name}"
+                )
+            continue
+
+        # Check if value is string before processing further
+        if isinstance(value, str):
+            result["text"].append({"name": key})
+            continue
+
+        # Unable to classify the field value
+        name = type(value).__name__
+        raise ValueError(
+            "Could not generate Redis index field type mapping "
+            + f"for metadata: '{key}': {name}"
+        )
+
+    return result
+
+
+def _prepare_metadata(metadata: Dict[str, Any]) -> Dict[str, Any]:
+    """
+    Prepare metadata for indexing in Redis by sanitizing its values.
+
+    - String, integer, and float values remain unchanged.
+    - None or empty values are replaced with empty strings.
+    - Lists/tuples of strings are joined into a single string with a comma separator.
+
+    Args:
+        metadata (Dict[str, Any]): A dictionary where keys are metadata
+            field names and values are the metadata values.
+
+    Returns:
+        Dict[str, Any]: A sanitized dictionary ready for indexing in Redis.
+
+    Raises:
+        ValueError: If any metadata value is not one of the known
+            types (string, int, float, or list of strings).
+    """
+
+    def raise_error(key: str, value: Any) -> None:
+        raise ValueError(
+            f"Metadata value for key '{key}' must be a string, int, "
+            + f"float, or list of strings. Got {type(value).__name__}"
+        )
+
+    clean_meta: Dict[str, Union[str, float, int]] = {}
+    for key, value in metadata.items():
+        if not value:
+            clean_meta[key] = ""
+            continue
+
+        # No transformation needed
+        if isinstance(value, (str, int, float)):
+            clean_meta[key] = value
+
+        # if it's a list/tuple of strings, we join it
+        elif isinstance(value, (list, tuple)):
+            if not value or isinstance(value[0], str):
+                clean_meta[key] = ",".join(value)
```

Comment:
Note that if there are tag values within the list that contain a comma, they will be split when indexed in RediSearch. Consider validating that there are no such values, or allowing the user to specify a different separator (API allows it).

Reply:
So currently, they could specify it in the schema and clean up the data themselves beforehand, but for the automatically generated metadata (how most will use it because they just use the data loaders), it's defaulting to ``,`` right now. Been thinking I should have a default here anyway. Good call out.